File: EjercicioGenerecidad3/src/main/java/com/mycompany/ejerciciogenerecidad3/EjercicioGenerecidad3.java
package com.mycompany.ejerciciogenerecidad3;
import java.util.*;
/**
 *
 * @author Vladimir Ohara
 */
class Catalogo<T> {
    private List<T> elementos;

    public Catalogo() {
        elementos = new ArrayList<>();
    }
//Inciso A
    public void agregar(T elemento) {
        elementos.add(elemento);
    }

    public boolean buscar(T elemento) {
        return elementos.contains(elemento);
    }

    public void mostrarTodos() {
        for (T elemento : elementos) {
            System.out.println(elemento);
        }
    }
}
class Libro {
    private String titulo;

    public Libro(String titulo) {
        this.titulo = titulo;
    }

    @Override
    public String toString() {
        return "Libro: " + titulo;
    }

    @Override
    public boolean equals(Object obj) {
        if (obj instanceof Libro) {
            return this.titulo.equals(((Libro) obj).titulo);
        }
        return false;
    }
}
class Producto {
    private String nombre;

    public Producto(String nombre) {
        this.nombre = nombre;
    }

    @Override
    public String toString() {
        return "Producto: " + nombre;
    }

    @Override
    public boolean equals(Object obj) {
        if (obj instanceof Producto) {
            return this.nombre.equals(((Producto) obj).nombre);
        }
        return false;
    }
}
public class EjercicioGenerecidad3 {

    public static void main(String[] args) {
        //El inciso B
       Catalogo<Libro> catalogoLibros = new Catalogo<>();
        catalogoLibros.agregar(new Libro("Cien Años de Soledad"));
        catalogoLibros.agregar(new Libro("El Principito"));

        System.out.println("Catálogo de Libros:");
        catalogoLibros.mostrarTodos();

       //El inciso C
        Catalogo<Producto> catalogoProductos = new Catalogo<>();
        catalogoProductos.agregar(new Producto("Laptop"));
        catalogoProductos.agregar(new Producto("Smartphone"));

        System.out.println("\nCatálogo de Productos:");
        catalogoProductos.mostrarTodos();
    }
}

File: EjercicioGenerecidad3/src/main/java/com/mycompany/ejerciciogenerecidad3/EjercicioGenerecidad3.py
from typing import Generic, TypeVar, List

T = TypeVar('T')

class Catalogo(Generic[T]):
    def __init__(self):
        self.elementos: List[T] = []

    def agregar(self, elemento: T):
        self.elementos.append(elemento)

    def buscar(self, elemento: T) -> bool:
        return elemento in self.elementos

    def mostrar_todos(self):
        for e in self.elementos:
            print(e)
class Libro:
    def __init__(self, titulo: str):
        self.titulo = titulo

    def __eq__(self, other):
        return isinstance(other, Libro) and self.titulo == other.titulo

    def __str__(self):
        return f"Libro: {self.titulo}"

class Producto:
    def __init__(self, nombre: str):
        self.nombre = nombre

    def __eq__(self, other):
        return isinstance(other, Producto) and self.nombre == other.nombre

    def __str__(self):
        return f"Producto: {self.nombre}"
if __name__ == "__main__":
    catalogo_libros = Catalogo[Libro]()
    catalogo_libros.agregar(Libro("Cien Años de Soledad"))
    catalogo_libros.agregar(Libro("El Principito"))

    print("Catálogo de libros:")
    catalogo_libros.mostrar_todos()

    catalogo_productos = Catalogo[Producto]()
    catalogo_productos.agregar(Producto("Laptop"))
    catalogo_productos.agregar(Producto("Smartphone"))

    print("\nCatálogo de productos:")
    catalogo_productos.mostrar_todos()
